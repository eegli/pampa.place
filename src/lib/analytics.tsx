import {useRouter} from 'next/router';
import Script from 'next/script';
import {useEffect} from 'react';
import {pageview} from './analytics-events';

export const initAnalytics = () => {
  const GA_TRACKING_ID = process.env.NEXT_PUBLIC_GA_ID;

  if (!GA_TRACKING_ID) {
    return () => null;
  }

  return function GoogleAnalytics() {
    const router = useRouter();

    useEffect(() => {
      const handleRouteChange = (url: string) => pageview(url);
      router.events.on('routeChangeComplete', handleRouteChange);
      return () => {
        router.events.off('routeChangeComplete', handleRouteChange);
      };
    }, [router.events]);

    return (
      <>
        <Script
          id="ga"
          strategy="afterInteractive"
          src={`https://www.googletagmanager.com/gtag/js?id=${GA_TRACKING_ID}`}
        />
        <Script
          id="ga-script"
          strategy="afterInteractive"
          dangerouslySetInnerHTML={{
            __html: `
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', '${GA_TRACKING_ID}');`,
          }}
        />
      </>
    );
  };
};
