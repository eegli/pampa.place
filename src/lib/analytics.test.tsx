/* DO use the original renderer */
import {render} from '@testing-library/react';
import {initAnalytics} from './analytics';

describe('Analytics initialization', () => {
  it('returns null if tracking ID is not set', () => {
    expect(process.env.NEXT_PUBLIC_GA_ID).toBeUndefined();
    const Script = initAnalytics();
    render(<Script />);
    expect(document.body).toMatchInlineSnapshot(`
      <body>
        <div />
      </body>
    `);
  });
  it('returns script tag if tracking ID is set', () => {
    process.env.NEXT_PUBLIC_GA_ID = 'test';
    expect(process.env.NEXT_PUBLIC_GA_ID).toBe('test');
    const Script = initAnalytics();
    render(<Script />);
    expect(document.body).toMatchInlineSnapshot(`
      <body>
        <div />
        <script
          data-nscript="afterInteractive"
          src="https://www.googletagmanager.com/gtag/js?id=test"
        />
        <script
          data-nscript="afterInteractive"
          id="ga-script"
        >
          
          window.dataLayer = window.dataLayer || [];
          function gtag(){dataLayer.push(arguments);}
          gtag('js', new Date());
          gtag('config', 'test', {
            page_path: window.location.pathname,
          });
        
        </script>
      </body>
    `);
  });
});
