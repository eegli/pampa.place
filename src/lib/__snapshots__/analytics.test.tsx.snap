// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics initialization returns null if tracking ID is not set 1`] = `
<body>
  <div />
</body>
`;

exports[`Analytics initialization returns script tag if tracking ID is set 1`] = `
<body>
  <div />
  <script
    data-nscript="afterInteractive"
    id="ga"
    src="https://www.googletagmanager.com/gtag/js?id=test-id"
  />
  <script
    data-nscript="afterInteractive"
    id="ga-script"
  >
    
            window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'test-id');
  </script>
</body>
`;
